modules/assim.sequential/R/downscale_function.R

```diff
@@ -1,67 +1,104 @@
+##' @param ensemble_data Character or list. Either a file path to an `.rds` file containing ensemble data, or an in-memory list where each element is a data.frame indexed by date.
+##' @param site_coordinates Character, data.frame, or sf object. Either:
+##'   - A file path to a `.csv` containing site coordinates with columns `"id"`, `"lat"`, and `"lon"`.
+##'   - A `data.frame` with the same structure.
+##'   - An `sf` object with point geometries.
+##' @param date Date. If SDA site run, format is yyyy/mm/dd; if NEON, yyyy-mm-dd. Restricted to years within the file or object supplied to 'ensemble_data'.
+##' @param carbon_pool Character. Carbon pool of interest. Name must match the carbon pool name found within the file or object supplied to 'ensemble_data'.
 ##' @details This function ensures that the specified date and carbon pool are present in the input data. It also checks the validity of the site coordinates and aligns the number of rows between site coordinates and carbon data.
 ##'
 ##' @description This function reads and checks the input data, ensuring that the required date and carbon pool exist, and that the site coordinates are valid.
 ##'
 ##' @return A list containing The read .rds data , The cleaned site coordinates, and the preprocessed carbon data.
 
-SDA_downscale_preprocess <- function(data_path, coords_path, date, carbon_pool) {
-  # Read the input data and site coordinates
-  input_data <- readRDS(data_path)
-  site_coordinates <- readr::read_csv(coords_path)
-  
+SDA_downscale_preprocess <- function(ensemble_data, site_coords, date, carbon_pool) {
+  # Read the input data and site coordinates (if inputs are file paths, load them; otherwise, use as provided)
+  if (is.character(ensemble_data)) {
+    input_data <- readRDS(ensemble_data)
+  } else {
+    input_data <- ensemble_data
+  }
+
+  if (is.character(site_coords)) {
+    site_coordinates <- readr::read_csv(site_coords)
+  } else {
+    site_coordinates <- site_coords
+  }
+
   # Convert input_data names to Date objects
   input_date_names <- lubridate::ymd(names(input_data))
   names(input_data) <- input_date_names
-  
+
   # Convert the input date to a Date object
   standard_date <- lubridate::ymd(date)
-  
+
   # Ensure the date exists in the input data
   if (!standard_date %in% input_date_names) {
     stop(paste("Date", date, "not found in the input data."))
   }
-  
+
   # Extract the carbon data for the specified focus year
   index <- which(input_date_names == standard_date)
   data <- input_data[[index]]
-  
+
   # Ensure the carbon pool exists in the input data
   if (!carbon_pool %in% names(data)) {
     stop(paste("Carbon pool", carbon_pool, "not found in the input data."))
   }
-  
+
   carbon_data <- as.data.frame(t(data[which(names(data) == carbon_pool)]))
   names(carbon_data) <- paste0("ensemble", seq(ncol(carbon_data)))
-  
+
+  # If sf object, convert to data.frame with 'lon' and 'lat'
+  if (inherits(site_coordinates, "sf")) {
```

Comment:
Curious why this is down here rather than up earlier where the other two site_coordinates cases were

Reply:
good point, no real reason